src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,684 @@
+  Striper::extent_to_file(m_cct, &m_dst_image_ctx->layout, m_dst_object_number,
+                          0, m_dst_image_ctx->layout.object_size, image_extents);
+
+  size_t total = 0;
+  for (auto &e : image_extents) {
+    std::map<object_t, std::vector<ObjectExtent>> src_object_extents;
+    Striper::file_to_extents(m_cct, m_src_image_ctx->format_string,
+                             &m_src_image_ctx->layout, e.first, e.second, 0,
+                             src_object_extents);
+    for (auto &p : src_object_extents) {
+      for (auto &s : p.second) {
+        m_src_objects.insert(s.objectno);
+        m_src_object_extents.push_back({s.objectno, s.offset, s.length});
+        total += s.length;
+      }
+    }
+  }
+
+  assert(total == m_dst_image_ctx->layout.object_size);
+
+  ldout(m_cct, 20) << m_src_object_extents.size() << " src extents" << dendl;
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_read_ops() {
+  librados::snap_t src_copy_point_snap_id = m_snap_map.rbegin()->first;
+  uint64_t prev_end_size = 0;
+  bool prev_exists = false;
+  librados::snap_t start_src_snap_id = 0;
+
+  m_read_ops = {};
+  m_read_snaps = {};
+  m_zero_interval = {};
+  m_end = {};
+
+  for (auto &pair : m_snap_map) {
+    assert(!pair.second.empty());
+    librados::snap_t end_src_snap_id = pair.first;
+    librados::snap_t end_dst_snap_id = pair.second.front();
+
+    interval_set<uint64_t> diff;
+    uint64_t end_size;
+    bool exists;
+    librados::snap_t clone_end_snap_id;
+    calc_snap_set_diff(m_cct, m_snap_set, start_src_snap_id,
+                       end_src_snap_id, &diff, &end_size, &exists,
+                       &clone_end_snap_id);
+
+    ldout(m_cct, 20) << "start_src_snap_id=" << start_src_snap_id << ", "
+                     << "end_src_snap_id=" << end_src_snap_id << ", "
+                     << "clone_end_snap_id=" << clone_end_snap_id << ", "
+                     << "end_dst_snap_id=" << end_dst_snap_id << ", "
+                     << "diff=" << diff << ", "
+                     << "end_size=" << end_size << ", "
+                     << "exists=" << exists << dendl;
+
+    if (exists) {
+      // clip diff to size of object (in case it was truncated)
+      if (end_size < prev_end_size) {
+        interval_set<uint64_t> trunc;
+        trunc.insert(end_size, prev_end_size);
+        trunc.intersection_of(diff);
+        diff.subtract(trunc);
+        ldout(m_cct, 20) << "clearing truncate diff: " << trunc << dendl;
+      }
+
+      // reads should be issued against the newest (existing) snapshot within
+      // the associated snapshot object clone. writes should be issued
+      // against the oldest snapshot in the snap_map.
+      assert(clone_end_snap_id >= end_src_snap_id);
+      if (clone_end_snap_id > src_copy_point_snap_id) {
+        // do not read past the copy point snapshot
+        clone_end_snap_id = src_copy_point_snap_id;
+      }
+
+      m_zero_interval[end_src_snap_id] = {};
```

Comment:
Thinking "out loud" because the ```m_end``` / ```m_dst_object_end``` / ```m_dst_zero_interval``` tracking seems a bit complicated, I wonder if instead of explicitly tracking holes and object end sizes, what if this is always initialized to the full extent of the object size and then valid data regions are removed? i.e. you would always be zeroed from ```end_size``` to object size (or zero to object size if ```!exists```), and for each snapshot you process, subtract out ```diff```. The next snapshot's zero interval starts w/ the previous snapshot's zero interval.

Reply:
@dillaman I might not see your idea, but the problems I saw trying something like you described:
- If the extent is initially initialized to zeros (instead of tracking zeros) I expect unnecessary zero op (when write is in the middle of the object, there will be also zero op from object start to this write).
- Before using `diff` to subtract out zeroes, I would need to transform it to destination offsets.
- Reading an src object may be restarted on `retry_missing_read`, and it adds some complication to store into intermediate data.

Anyway, I was able to get rid off `m_end` / `m_dst_object_end`, and I suppose it is a little simpler now. Could you please take a look?

Reply:
> - If the extent is initially initialized to zeros (instead of tracking zeros) I expect unnecessary zero op (when write is in the middle of the object, there will be also zero op from object start to this write).
After computing and executing all reads, you could trim off any write ops associated before the first read (by snap_id). 

> - Before using `diff` to subtract out zeroes, I would need to transform it to destination offsets.
You should have to do that regardless since all writes need to be translated to the destination object. Also, you could either track the holes by destination extents or you could have a mapping from source object number to source extents and then combine all source extents prior to the write step.

> - Reading an src object may be restarted on `retry_missing_read`, and it adds some complication to store into intermediate data.
Hmm -- on restart it would re-zero the destination extents associated w/ that source object and then when iterating over each src_snap_id, it would base the updates on the previous state's map. If tracking by zeroes by destination extents, you would have to intersect the source extents to ensure you only update appropriate section of the interval_set.

Regardless, if you were able to eliminate ```m_end``` and ```m_dst_object_end```, perhaps it's good-to-go as-is. Let me look.